test_progs/pcq.c

```diff
@@ -0,0 +1,1066 @@
+	return rc;
+}
+
+/* XXX Move this to famfs_lib_util.c or some such - it is now shared with the cli */
+static s64 get_multiplier(const char *endptr)
+{
+	size_t multiplier = 1;
+
+	if (!endptr)
+		return 1;
+
+	switch (*endptr) {
+	case 'k':
+	case 'K':
+		multiplier = 1024;
+		break;
+	case 'm':
+	case 'M':
+		multiplier = 1024 * 1024;
+		break;
+	case 'g':
+	case 'G':
+		multiplier = 1024 * 1024 * 1024;
+		break;
+	case 0:
+		return 1;
+	}
+	++endptr;
+	if (*endptr) /* If the unit was not the last char in string, it's an error */
+		return -1;
+	return multiplier;
+}
+
+void
+pcq_usage(int   argc,
+	    char *argv[])
+{
+	char *progname = argv[0];
+
+	printf("\n"
+	       "pcq: Run a producer/consumer queue test\n"
+	       "\n"
+	       "This test creates a set of files to use as a producer/consumer queue\n"
+	       "and sends messages through the queue. You can run one copy of this program\n"
+	       "that does both the producer & consumer functions, but to test shared memory\n"
+	       "you need to run one copy of this program as producer an a second copy (on a\n"
+	       "different node) as consumer.\n"
+	       "\n"
+	       "A producer/consumer queue is implemented as a set of files. The producer file\n"
+	       "contains the producer index and the buckets. The consumer file contains the\n"
+	       "consumer index. The producer maps the producer file writable and the consumer\n"
+	       "file read-only; the consumer does the opposite.\n"
+	       "\n"
+	       "EXAMPLES:\n"
+	       "Create a producer/consumer queue with 4096 buckets which are 1K each:\n"
+	       "    %s --create --bsize 1024 --nbuckets 4K <queuename>\n"
+	       "\n"
+	       "Just run a producer:\n"
+	       "    %s --producer [Args] /mnt/famfs/<queuename>\n"
+	       "\n"
+	       "Run a consumer:\n"
+	       "    %s --consumer [Args] /mnt/famfs/<queuename>\n"
+	       "\n"
+	       "Run a producer and a consumer from a single process:\n"
+	       "    %s --producer --consumer [Args] /mnt/famfs/<queuename>\n"
+	       "\n"
+	       "Drain a pcq\n"
+	       "    %s --drain [Args] /mnt/famfs/<queuename>\n"
+	       "\n"
+	       "Check the state of a producer/consumer queue (maps both fies read-only:\n"
+	       "    %s --info [Args] /mnt/famfs/<queuename>\n"
+	       "\n"
+	       "Arguments:\n"
+	       "\n"
+	       "Queue Creation:\n"
+	       "    -C|--create               - Create a producer/consumer queue\n"
+	       "    -b|--bsize <bucketsize>   - size of messages including sequence number\n"
+	       "                                and crc (ignored if queue already exists)\n"
+	       "    -n|--nbuckets <nnbuckets> - Number of buckets in the queue\n"
+	       "                                (ignored if queue already exists)\n"
+	       "\n"
+	       "Running producers and consumers:"
+	       "    -N|--nmessages <n>        - Number of messages to send and/or receive\n"
+	       "    -t|--time <seconds>       - Run for the specified duration\n"
```

Comment:
If neither the time option or the number of message option is given, it runs indefinitely (which is fine) but to stop it we need to do Ctrl+C and we don't get the number of messages transferred and other info.

One suggestion is to add a default timeout value which is kicked in in case -t or -N option are not given.

Reply:
Good catch. I think there should be an option to run indefinitely, but "--time 0" seems better for that. Oh, actually it already works that way because "runtime = nmessages = 0" is the default, which runs the threads in "nmessages" mode with a target value of 0 which is never a match. 

Will think on this...